test/cri-containerd/clone_test.go

```diff
@@ -0,0 +1,441 @@
+
+	// create multiple clones
+	for i := 0; i < nClones; i++ {
+		createClonedContainer(ctx, t, client, templatePodID, templateContainerID, i, &clonedPodIDs[i], &clonedContainerIDs[i])
+	}
+
+	for i := 0; i < nClones; i++ {
+		verifyContainerExec(ctx, t, client, clonedContainerIDs[i])
+	}
+}
+
+// Test if a normal container can be created inside a clond pod alongside the cloned
+// container.
+// TODO(ambarve): This doesn't work as of now. Enable this test when the bug is fixed.
+func DisabledTest_NormalContainerInClonedPod_WCOW(t *testing.T) {
+	requireFeatures(t, featureWCOWHypervisor)
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	client := newTestRuntimeClient(t)
+	var templatePodID, templateContainerID string
+	var clonedPodID, clonedContainerID string
+	var stdContainerID string
+	pullRequiredImages(t, []string{imageWindowsNanoserver})
+
+	defer cleanupPod(t, client, ctx, &templatePodID)
+	defer cleanupContainer(t, client, ctx, &templateContainerID)
+	defer cleanupPod(t, client, ctx, &clonedPodID)
+	defer cleanupContainer(t, client, ctx, &clonedContainerID)
+	defer cleanupContainer(t, client, ctx, &stdContainerID)
+
+	// create template pod & container
+	createTemplateContainer(ctx, t, client, getTemplatePodConfig("templatepod"), getTemplateContainerConfig("templatecontainer"), &templatePodID, &templateContainerID)
+	verifyTemplateContainerState(ctx, t, client, templateContainerID)
+
+	// create a cloned pod and a cloned container
+	cloneSandboxRequest := getClonedPodConfig(1, templatePodID)
+	cloneContainerRequest := getClonedContainerConfig(1, templateContainerID)
+	createPodAndContainer(ctx, t, client, cloneSandboxRequest, cloneContainerRequest, &clonedPodID, &clonedContainerID)
+
+	// create a normal container in cloned pod
+	stdContainerRequest := getStandardContainerConfig("standard-container")
+	stdContainerRequest.PodSandboxId = clonedPodID
+	stdContainerRequest.SandboxConfig = cloneSandboxRequest.Config
+	stdContainerID = createContainer(t, client, ctx, stdContainerRequest)
+	startContainer(t, client, ctx, stdContainerID)
+
+	verifyContainerExec(ctx, t, client, clonedContainerID)
+	verifyContainerExec(ctx, t, client, stdContainerID)
+}
+
+// A test for cloning multiple pods first and then cloning one container in each
+// of those pods.
+func Test_CloneContainersWithClonedPodPool_WCOW(t *testing.T) {
+	requireFeatures(t, featureWCOWHypervisor)
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	client := newTestRuntimeClient(t)
+	nClones := 3
+	var templatePodID, templateContainerID string
+	clonedPodIDs := make([]string, nClones)
+	clonedContainerIDs := make([]string, nClones)
+
+	defer cleanupPod(t, client, ctx, &templatePodID)
+	defer cleanupContainer(t, client, ctx, &templateContainerID)
+	for i := 0; i < nClones; i++ {
+		defer cleanupPod(t, client, ctx, &clonedPodIDs[i])
+		defer cleanupContainer(t, client, ctx, &clonedContainerIDs[i])
+	}
```

Comment:
Why can't we just put these immediately after creating the pod/containers later in the test? Not sure I like this setup.

Reply:
We must defer these functions before we start creation of pods/containers. In case there is a failure in the `createTemplateContainer` function we won't reach the line next to it which sets up a defer function to cleanup the pod/container created in that function. Setting these function ahead will make sure that they are run even in case of a failure. That is why we have to pass the pointers to the podID and containerID variables.